### PR TITLE
Library + compiler improvements.

### PR DESCRIPTION
**Pull request by ryzhyk**
See commits for details.

### CODE REVIEW

File: test/datalog_tests/std_test.dump.expected [hunk 1]
```diff
@@ -22,3 +22,15 @@ std_test::Pow{.descr = "$pow32(2: u64, 32)", .val = "4294967296"}
 std_test::Pow{.descr = "pow32(2: u32, 32)", .val = "0"}
 std_test::Default{.descr = "DefaultEnum", .val = "DEnum1"}
 std_test::Default{.descr = "u32", .val = "0"}
+std_test::MapDelete{.descr = "alphabet", .m = [(0, "a"), (1, "b"), (2, "c"), (3, "d"), (4, "e")], .s = std::None{}}
+std_test::MapDelete{.descr = "alphabet\\4", .m = [(0, "a"), (1, "b"), (2, "c"), (3, "d")], .s = std::Some{.x = "e"}}
```
Comment (mbudiu-vmw) [inline test/datalog_tests/std_test.dump.expected:26]: did you intend to have this double slash here?

File: test/datalog_tests/std_test.dl [hunk 2]
```diff
@@ -14,6 +14,32 @@ SortedVectorInPlace(sorted) :-
         v2
     }.
 
+output relation IntVecTest(descr: string, vec: Vec<s64>)
+
+IntVecTest("vec0", vec0()).
+IntVecTest("vec0.resize(10)", {var v = vec0(); v.resize(10, -100); v}).
```
Comment (mbudiu-vmw) [inline test/datalog_tests/std_test.dl:20]: should the string be resize(10, -100)?

File: lib/std.dl [hunk 3]
```diff
@@ -497,14 +497,32 @@ function to_set(s: Vec<'A>): Set<'A> {
...
+/* Resize vector.  If `new_size` is greater than `v.len()`, then
+ * all new slots are filled with `value`. */
+function resize(v: mut Vec<'X>, new_len: usize, value: 'X) {
+    vec_resize(v, new_len, value)
+}
+
+/* Swap value at index `idx` for `value`.
+ * Stores the old value at this index in `value` unless `idx` exceeds
+ * the size of the vector (in which case the vector remains unmodified). */
+function swap_nth(v: mut Vec<'X>, idx: usize, value: mut 'X) {
```
[the remark this answers is not in the document]
Comment (ryzhyk) [reply lib/std.dl:517]: good idea

File: src/Language/DifferentialDatalog/NS.hs [hunk 4]
```diff
@@ -142,12 +142,24 @@ getRelation d n = fromJust $ lookupRelation d n
...
+-- This function is _unsafe_ to use before type inference.
 ctxVars :: DatalogProgram -> ECtx -> ([Var], [Var])
-ctxVars d ctx =
+ctxVars d ctx = ctxVars' d ctx True
+
+-- All variables available in the scope: (l-vars, read-only vars).
+--
+-- The 'with_types' flag is true if 'd' contains enough type information
+-- to determine variable types (i.e., it has been through type inference).
+-- When false, `ctxVars` may misclassify
```
[the remark this answers is not in the document]
Comment (ryzhyk) [reply src/Language/DifferentialDatalog/NS.hs:160]: I clarified the contract.  This function is private to the module and is only called in a safe way there.

File: src/Language/DifferentialDatalog/Parse.hs [hunk 5]
```diff
@@ -297,13 +297,13 @@ typeDef = (TypeDef nopos []) <$ reserved "typedef" <*> typeIdent <*>
...
                          <*> (parens $ commaSep farg)
-                         <*> (colon *> typeSpecSimple)
+                         <*> (option (TTuple nopos []) (colon *> typeSpecSimple))
                          <*> (return Nothing))
        <|>
        (Function nopos [] <$  reserved "function"
                          <*> funcIdent
                          <*> (parens $ commaSep farg)
-                         <*> (colon *> typeSpecSimple)
+                         <*> (option (TTuple nopos []) (colon *> typeSpecSimple))
```
[the remark this answers is not in the document]
Comment (ryzhyk) [reply src/Language/DifferentialDatalog/Parse.hs:306]: That's the right thing to do, but we don't have the infrastructure for warning atm.  In particular, it should be possible to deprecate warnings or convert them to errors.

File: test/datalog_tests/json_test.dl [hunk 6]
```diff
@@ -204,3 +204,34 @@ JsonTestValue("wrapped " ++ wenum2(),
...
+    // convert null into map
+    jval.set_by_ptr([JKeyPtr{i"nested"}, JKeyPtr{i"z"}, JIdxPtr{9}, JKeyPtr{i"q"}], from_json_string(struct_with_map1()).unwrap_or_default());
+
+    // this should be a no-op.
+    jval.set_by_ptr([JKeyPtr{i"t"}, JKeyPtr{i"z"}], from_json_string(struct_with_map1()).unwrap_or_default());
+
+    jval
+}
+
+JsonTest("get_by_ptr(nested/z/10/b)", mutilate_jval().get_by_ptr([JKeyPtr{i"nested"}, JKeyPtr{i"z"}, JIdxPtr{10}, JKeyPtr{i"b"}]).to_json_string().unwrap_or_default()).
```
[the remark this answers is not in the document]
Comment (ryzhyk) [reply test/datalog_tests/json_test.dl:235]: The only use-case I have uses non-standard syntax for pointers, so I am taking the lazy approach here.

**Pull request merged**